packages/task/src/browser/task-configurations.ts

```diff
@@ -0,0 +1,257 @@
+                })
+                .catch(err => {
+                    reject(err);
+                });
+        });
+    }
+
+    protected getConfigFile(rootDir: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            // compute URL of the tasks file to watch
+            const cfgFile = new URI(rootDir).resolve(this.TASKFILEPATH).resolve(this.TASKFILE).toString();
+            this.fileSystem.exists(cfgFile).then(exists => {
+                if (exists) {
+                    resolve(cfgFile);
+                } else {
+                    this.logger.error(`config file tasks.json does not exist under ${rootDir}`);
+                    reject(`config file does not exist under: ${rootDir}`);
+                }
+            });
+        });
+    }
+
+    getTaskLabels(): string[] {
+        return [...this.tasksMap.keys()];
+    }
+
+    getTask(taskLabel: string): Promise<TaskOptions> {
+        return new Promise((resolve, reject) => {
+            if (this.tasksMap.has(taskLabel)) {
+                resolve(this.tasksMap.get(taskLabel));
+            } else {
+                reject(`Task with label ${taskLabel} not found`);
+            }
+        });
+    }
+
+    protected watchConfigFile(uri: string): Promise<number> {
+        // watch file for changes
+        return new Promise(resolve => {
+            this.watcherServer.watchFileChanges(uri)
+                .then(id => {
+                    this.toDispose.push(Disposable.create(() =>
+                        this.watcherServer.unwatchFileChanges(id))
+                    );
+                    resolve(id);
+                });
+        });
+    }
+
+    // a task config file we're watching has changed - update task list
+    protected onDidTaskFileChange(changes: DidFilesChangedParams): Promise<void> {
+        return new Promise((resolve, reject) => {
+            // one tasks.json file we're watching has changed?
+            if (this.hasWatchedFileChanged(changes.changes)) {
+                this.tasksMap.clear();
+                const promises: Promise<TaskOptions[]>[] = [];
+
+                // re-read all config files
+                for (const fileUri of this.watchedConfigFileUris) {
+                    promises.push(this.readTasks(fileUri));
+                }
+
+                Promise.all(promises)
+                    .then(tasksArrayArray => {
+                        for (const taskArray of tasksArrayArray) {
+                            for (const task of taskArray) {
+                                this.tasksMap.set(task.label, task);
+                            }
+                        }
+                        resolve();
+                    })
+                    .catch(err => {
+                        this.logger.error(err);
+                        reject(err);
+                    });
+            }
+        });
+    }
+
+    protected hasWatchedFileChanged(changes: FileChange[]): boolean {
```

Comment:
Just a marker for @akosyakov  #942

Reply:
You should not use the watcher server in the frontend, it is the impl details of `FileSystemWatcher`, the watcher takes care of servers and reconnection to them if a connection is down as well as preserving watching.


I am working on PR which will add an optional callback to `FileSystemWatcher.watch(uri): Promise<Disposable>` that you can pass it to avoid dispatching changes to other clients.

Reply:
@akosyakov  ok, I will update this code accordingly, when your PR is ready.